### PR TITLE
Create AwareModel.php

### PR DESCRIPTION
**Pull request by grillkoning**
Pull Request for Issue #21867.

### Summary of Changes
to make joomla's admin model aware of subforms so you can actually handle them in seperate tables with their corresponding model

### Testing Instructions
use a subform to add a row

### Expected result
some DB-table has a row added which is handled by default CMS functions

### Actual result
the row gets stashed away in a json field making it unpractical for sorting, searching, ACL, and just about anything

### Documentation Changes Required
tbd

### Example usage
Instead of extending \Joomla\CMS\MVC\ModelAdminModel ; extend this class to have your form handle subforms with their own respective models and forms

Off course you'll need to provide the model and form you want your subform to use. So in the example below you'd provide:

```
/**
 * Model for a list of History items
 *
 * @since 0.0.1
 */
class AmberModelHistories extends ListModel
{
// as you would when creating a listview like normal 
}
```

snippet from example form .xml:
```
<field
	name="histories"
	type="subform"
	formsource="/administrator/components/com_amber/models/forms/histories.xml"
	multiple="true"
	layout="joomla.form.field.subform.repeatable-table"
	/>
```
content of histories.xml:
note the fieldgroup name which needs to match the name of the subform
```
<?xml version="1.0" encoding="utf-8"?>
<form>
<fieldgroup name="histories">
		<field
				name="historyID"
				type="hidden"
				/>
		<field
				name="resourceID"
				type="resource"
				label="COM_AMBER_RESOURCE_TITLE_LABEL"
				description="COM_AMBER_RESOURCE_TITLE_DESC"
				class="inputbox"
				/>
		<field
				name="difQty"
				type="number"
				label="COM_AMBER_QUANTITY_LABEL"
				description="COM_AMBER_QUANTITY_DESC"
				class="inputbox"
				/>
</fieldgroup>
				
</form>
```
the controller for the subform:
/administrator/compontents/com_amber/controllers/histories.php

```
class AmberControllerHistories extends AdminController
{

	public function getModel($name = 'History', $prefix = 'AmberModel', $config = array('ignore_request' => true))
	{
		$model = parent::getModel($name, $prefix, $config);
                return $model;
	}

```

### CODE REVIEW

**Comment by ghost** [comment]
@grillkoning i changed your Comment to be more readable and linked to Issue.

File: libraries/src/MVC/Model/AwareModel.php [hunk 1]
```diff
@@ -0,0 +1,335 @@
+<?php
+/**
+ *
+ * @subpackage amber
+ *
+ * @copyright Copyright (C) 2005 - 2018 Chris Rutten. All rights reserved.
```
Comment (jeckodevelopment) [inline libraries/src/MVC/Model/AwareModel.php]: Generally files in the CMS (unless they are from vendor) have this copyright note:

`@copyright  Copyright (C) 2005 - 2018 Open Source Matters, Inc. All rights reserved.`

File: libraries/src/MVC/Model/AwareModel.php [hunk 2]
```diff
@@ -0,0 +1,335 @@
+<?php
+/**
+ *
+ * @subpackage amber
```
Comment (jeckodevelopment) [inline libraries/src/MVC/Model/AwareModel.php]: Where does this subpackage name comes from?

File: libraries/src/MVC/Model/AwareModel.php [hunk 3]
```diff
@@ -0,0 +1,335 @@
+<?php
+/**
+ *
```
Comment (jeckodevelopment) [inline libraries/src/MVC/Model/AwareModel.php:4]: Generally CMS files have this note at row 3:

` * Joomla! Content Management System`

File: libraries/src/MVC/Model/AwareModel.php [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+ * @subpackage amber
+ *
+ * @copyright Copyright (C) 2005 - 2018 Chris Rutten. All rights reserved.
+ * @license GNU General Public License version 2 or later; see LICENSE.txt
+ *
+ *          Purpose: to make joomla's admin model aware of subforms so you can
```
Comment (jeckodevelopment) [inline libraries/src/MVC/Model/AwareModel.php]: Joomla with capital J

**Review by jeckodevelopment — CHANGES_REQUESTED** [review]

File: libraries/src/MVC/Model/AwareModel.php (hunk 2 above)
Comment (zero-24) [reply libraries/src/MVC/Model/AwareModel.php]: subpackage is not used anymore in the CMS codebase but was used in the past so can be removed

**Comment by zero-24** [comment]
@grillkoning please take a look into the drone / PHPCS reports: http://ci.joomla.org/joomla/joomla-cms/9814/4

File: libraries/src/MVC/Model/AwareModel.php (hunk 2 above)
Comment (brianteeman) [reply libraries/src/MVC/Model/AwareModel.php]: @zero-24 are you sure about that - i think you are confused :)

Comment (zero-24) [reply libraries/src/MVC/Model/AwareModel.php]: There are no subpackage tags in namespaced classes see:
https://github.com/joomla/joomla-cms/blob/920c42921c72745746d1537ca97027a27efdfca3/libraries/src/MVC/Model/AdminModel.php#L1-L22

File: libraries/src/MVC/Model/AwareModel.php (hunk 4 above)
Comment (zero-24) [inline libraries/src/MVC/Model/AwareModel.php]: I would actually move this content to a docs page.

File: libraries/src/MVC/Model/AwareModel.php (hunk 2 above)
Comment (brianteeman) [reply libraries/src/MVC/Model/AwareModel.php]: ah i see - just in components etc not libraries

**Comment by laoneo** [comment]
Can you extend the description of this pr what you have in mind with it and a code example? It will be then easier to review.

**Comment by grillkoning** [comment]
thanks to all of you for your comments and attention. hopefully this will grow to be a useful feature to many.

File: libraries/src/MVC/Model/AwareModel.php (hunk 3 above)
Comment (grillkoning) [reply libraries/src/MVC/Model/AwareModel.php:4]: @jeckodevelopment thanks for the pointer; I have restyled the code

**Review by grillkoning — COMMENTED** [review]
thanks for the pointer; I have restyled the code

File: libraries/src/MVC/Model/AwareModel.php [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+ */
+namespace Joomla\CMS\MVC\Model;
+defined('_JEXEC') or die('Restricted access');
+use Joomla\CMS\Table\Table;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\Form;
+/**
+ * Extends AdminModel to be aware of subforms
+ *
+ * @since  3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
 * @since  __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 6]
```diff
@@ -0,0 +1,365 @@
+<?php
+/**
+ * Joomla! Content Management System
+ *
+ * @copyright  Copyright (C) 2005 - 2018 Open Source Matters, Inc. All rights reserved.
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
 * @copyright  (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+ * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ */
+namespace Joomla\CMS\MVC\Model;
+defined('_JEXEC') or die('Restricted access');
+use Joomla\CMS\Table\Table;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\Form;
+/**
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
namespace Joomla\CMS\MVC\Model;

defined('_JEXEC') or die('Restricted access');

use Joomla\CMS\Table\Table;
use Joomla\CMS\Factory;
use Joomla\CMS\Form\Form;

/**
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+ *
+ */
+class AwareModel extends AdminModel
+{
+	/**
+	 * An array of Joomla! Form objects
+	 * Joomla\CMS\Form\Form
+	 *
+	 * @var    array($pluralName)
+	 * @since  3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+	 */
+	private $subforms = null;
+	
+	/**
+	 * An array of Joomla! table objects
+	 * Joomla\CMS\Table\Table
+	 *
+	 * @var    array($pluralName)
+	 * @since  3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+	 */
+	private $subtables = null;
+	
+	/*
+	 * An array of Joomla! models of the type AdminModel
+	 * Joomla\CMS\MVC\Model\AdminModel
+	 *
+	 * @var    array($pluralName)
+	 * @since  3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+	 */
+	private $submodels = null;
+	
+	/**
+	 * The prefix to use when loading submodels.
+	 *
+	 * @var    string  
+	 * @since  3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+	
+	/**
+	 * Have the parent save the base Form and afterwards
+	 * save the subforms by ourselves.
+	 *
+	 * @param   array  $data  The form data.
+	 * 
+	 * @return  boolean  True on success, False on error.
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+		return $result;
+	}
+	/**
+	 * Retreive a certain subTable
+	 *
+	 * @param   string  $name  The name of the subtable to retrieve
+	 * 
+	 * @return  Table
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+		return $this->getSubTables()[$name];
+	}
+	/**
+	 * Retreive a certain subModel
+	 *
+	 * @param   string  $name  The name of the submodel to retrieve
+	 * 
+	 * @return  AdminModel
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+		return $this->getSubModels()[$name];
+	}
+	/**
+	 * Retrieve a certain subForm
+	 *
+	 * @param   string  $name  The name of the subForm to retrieve
+	 * 
+	 * @return  Form
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+	}
+	/**
+	 * Compare the posted form with the previously served form so see if anything
+	 * has been deleted by the user in the meantime
+	 *
+	 * @param   array  $data  The form data.
+	 *
+	 * @return  void
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+	}
+	/**
+	 * traverse the available subtables to see which parts of the given $data to
+	 * store in which $table
+	 *
+	 * @param   array  $data  The form data.
+	 *
+	 * @return  boolean  True on success, False on error.
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+				}
+			}
+		}
+	}
+	/**
+	 * Retrieve an array with all submodels in this object
+	 *
+	 * @return  array  An array of Joomla models
+	 * 
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+			$this->loadSubForms();
+		}
+		return $this->submodels;
+	}
+	/**
+	 * Loads all the subforms in this form object
+	 *
+	 * @return  void
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+			}
+		}
+	}
+	/**
+	 * Retrieve the prefix to use when loading forms or tables
+	 * Guessing one if none has been set.
+	 *
+	 * @return  string
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 21]
```diff
@@ -0,0 +1,365 @@
...
+		return ($this->prefix);
+	}
+	/**
+	 * Sets the prefix to use when loading tables or forms
+	 *
+	 * @param   string  $prefix  The prefix to use when loading tables or forms
+	 * 
+	 * @return  void
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+	public function setPrefix (string $prefix)
+	{
+		$this->prefix = $prefix;
+	}
+	/**
+	 * Retrieves a certain subform
+	 *
+	 * @return  Form
+	 * 
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+			$this->loadSubForms();
+		}
+		return $this->subforms;
+	}
+	/**
+	 * Loads the subtables into the object
+	 *
+	 * @return  void
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 24]
```diff
@@ -0,0 +1,365 @@
...
+			$controllerclass = ucfirst($prefix) . 'Controller' . ucfirst($pluralname);
+			$filename = JPATH_COMPONENT_ADMINISTRATOR . '/controllers/' . $pluralname . '.php';
+			require_once $filename;
+			$controller = new $controllerclass;
+			$singularname = $controller->getModel()->get('name');
+			$this->subtables[$pluralname] = $controller->getModel()->getTable($singularname);
+		}
+	}
+	/**
+	 * Retreives an array with all the subforms in this form
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * Retrieves an array with all the subforms in this form
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+	 *
+	 * @return  array($pluralName)
+	 *
+	 * @since   3.9
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+	 * @param   string   $name     The name of the form.
+	 * @param   string   $source   The form source. Can be XML string if file flag is set to false.
+	 * @param   array    $options  Optional array of options for the form creation.
+	 * @param   boolean  $clear    Optional argument to force load a new form.
+	 * @param   string   $xpath    An optional xpath to search for the fields.
+	 *
+	 * @return  Form|boolean  Form object on success, false on error.
+	 *
+	 * @see     \Joomla\CMS\Form\Form
+	 * @since   1.6
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/MVC/Model/AwareModel.php [hunk 27]
```diff
@@ -0,0 +1,365 @@
...
+	}
+	/**
+	 * Abstract method for getting the form from the model.
+	 *
+	 * @param   array    $data      Data for the form.
+	 * @param   boolean  $loadData  True if the form is to load its own data (default case), false if not.
+	 *
+	 * @return  Form|boolean  A Form object on success, false on failure
+	 *
+	 * @since   1.6
```
Comment (PhilETaylor) [inline libraries/src/MVC/Model/AwareModel.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Comment by laoneo** [comment]
Sorry it took so long to respond. As this is a nice addition for the MVC API, it is a new feature which should be adapted Joomla 4. Can you rebase the pr to the 4.2-dev branch? In the meantime I'm closing the pr, when ready please reopen again so we can leave feedback. Thank you very much for your contribution and help making Joomla better.

**Pull request closed without merging**